crates/wasmtime/src/runtime/debug.rs

```diff
@@ -1,172 +1,363 @@
+    fp: usize,
+    /// The resolved module handle for the physical PC.
+    ///
+    /// The module for each inlined frame within the physical frame is
+    /// resolved from the vmctx reachable for each such frame; this
+    /// module isused only for looking up the frame table.
+    module: Module,
+    /// The Wasm PC for this frame.
+    wasm_pc: u32,
+    /// The frame descriptor for this frame.
+    frame_descriptor: FrameTableDescriptorIndex,
+    /// The stack shape for this frame.
+    stack_shape: FrameStackShape,
 }
 
-fn relocate_dwarf_sections(bytes: &mut [u8], code_region: (*const u8, usize)) -> Result<(), Error> {
-    let mut relocations = Vec::new();
-    let obj = File::parse(&bytes[..]).map_err(obj::ObjectCrateErrorWrapper)?;
-    for section in obj.sections() {
-        let section_start = match section.file_range() {
-            Some((start, _)) => start,
-            None => continue,
+impl VirtualFrame {
+    /// Return virtual frames corresponding to a physical frame, from
+    /// outermost to innermost.
+    fn decode(store: &StoreOpaque, frame: Frame, is_trapping_frame: bool) -> Vec<VirtualFrame> {
+        let module = store
+            .modules()
+            .lookup_module_by_pc(frame.pc())
+            .expect("Wasm frame PC does not correspond to a module");
+        let base = module.code_object().code_memory().text().as_ptr() as usize;
+        let pc = frame.pc().wrapping_sub(base);
+        let table = module.frame_table();
+        let pc = u32::try_from(pc).expect("PC offset too large");
+        let pos = if is_trapping_frame {
+            FrameInstPos::Pre
+        } else {
+            FrameInstPos::Post
         };
-        for (off, r) in section.relocations() {
-            if r.kind() != RelocationKind::Absolute
-                || r.encoding() != RelocationEncoding::Generic
-                || r.size() != 64
-            {
-                continue;
-            }
-
-            let sym = match r.target() {
-                RelocationTarget::Symbol(index) => match obj.symbol_by_index(index) {
-                    Ok(sym) => sym,
-                    Err(_) => continue,
-                },
-                _ => continue,
-            };
-            relocations.push((
-                section_start + off,
-                (code_region.0 as u64)
-                    .wrapping_add(sym.address())
-                    .wrapping_add(r.addend() as u64),
-            ));
+        let Some(program_points) = table.find_program_point(pc, pos) else {
+            return vec![];
+        };
+
+        program_points
+            .map(|(wasm_pc, frame_descriptor, stack_shape)| VirtualFrame {
+                fp: frame.fp(),
+                module: module.clone(),
+                wasm_pc,
+                frame_descriptor,
+                stack_shape,
+            })
+            .collect()
+    }
+}
+
+/// A view of a frame that can decode values in that frame.
+pub struct FrameView {
+    slot_addr: usize,
```

Comment:
I think this is going to have to change one way or another to make this sound. Some things I can think of to break soundness with this are:

* I think a `FrameView` could get paired with the wrong `Store`.
* A `FrameView` could be persisted outside of a stack trace and then later read once the frames are all gone (or with the wrong store)

Personally I think it would be worth trying to consolidate the `unsafe` here in this module as much as possible. Ideally we'd just have one tiny "thing", whatever that is, which does the `unsafe` bits of actually reading from the stack and, ideally, we'd be able to read the definition of this "thing" and trivially conclude "yep that's fine".

Reply:
Fixed in 2df6a9140c per the sketch below.